provider/datagen/src/driver.rs

```diff
@@ -11,14 +11,230 @@ use icu_locid_transform::LocaleFallbacker;
+    /// Stylized on the CLI as: "full"
+    pub fn full() -> Self {
+        Self {
+            langid: None,
+            include_ancestors: false,
+            include_descendants: true,
+        }
+    }
+}
+
+impl fmt::Display for LocaleFamily {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match (
+            &self.langid,
+            self.include_ancestors,
+            self.include_descendants,
+        ) {
+            (Some(langid), true, true) => write!(f, "{}", langid),
+            (Some(langid), true, false) => write!(f, "^{}", langid),
+            (Some(langid), false, false) => write!(f, "@{}", langid),
+            (Some(_), false, true) => unreachable!(),
+            (None, _, _) => write!(f, "full"),
+        }
+    }
+}
+
+/// Options bag configuring locale inclusion and behavior when runtime fallback is enabled.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Default)]
+#[non_exhaustive]
+pub struct FallbackOptions {
+    /// The location in code where fallback will be performed at runtime.
+    ///
+    /// If not set, determined by the exporter: internal fallback is used if the exporter
+    /// supports internal fallback.
+    pub runtime_fallback_location: Option<RuntimeFallbackLocation>,
+    /// The aggressiveness of deduplication of data payloads.
+    ///
+    /// If not set, determined by `runtime_fallback_location`. If internal fallback is enabled,
+    /// a more aggressive deduplication strategy is used.
+    pub deduplication_strategy: Option<DeduplicationStrategy>,
+}
+
+#[derive(Debug, Clone)]
+enum LocalesWithOrWithoutFallback {
+    WithFallback {
+        locales: HashSet<LocaleFamily>,
+        options: FallbackOptions,
+    },
+    WithoutFallback {
+        locales: HashSet<LanguageIdentifier>,
+    },
+}
+
+impl fmt::Display for LocalesWithOrWithoutFallback {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::WithFallback { locales, options } => {
+                let mut sorted_locales =
+                    locales.iter().map(ToString::to_string).collect::<Vec<_>>();
+                sorted_locales.sort();
+                let start = match options.runtime_fallback_location {
+                    None => unreachable!(),
+                    Some(RuntimeFallbackLocation::Internal) => "with internal fallback",
+                    Some(RuntimeFallbackLocation::External) => "with external fallback",
+                };
+                write!(f, "{start}, ")?;
+                match options.deduplication_strategy {
+                    Some(x) => x.describe(f)?,
+                    None => unreachable!(),
+                }
+                write!(f, ", and these locales: {:?}", sorted_locales)
+            }
+            Self::WithoutFallback { locales } => {
+                let mut sorted_locales =
+                    locales.iter().map(ToString::to_string).collect::<Vec<_>>();
+                sorted_locales.sort();
```

Comment:
Found something cool for this: https://internals.rust-lang.org/t/add-a-vec-sorted-function/11948/3. We already have `itertools` in datagen

Reply:
Note that it returns `std::vec::IntoIter`, so it still does the same thing under the hood, just a bit nicer syntax and a bit more prone to error because if you call any other iterator methods like `.map()` then you end up double-allocating. It also seems a bit risky that `std::vec::IntoIter::collect` uses specialization. But I can do this if you like.

Reply:
not blocking